app/celery/nightly_tasks.py

```diff
@@ -409,3 +412,48 @@ def pick_bst_quarter_to_delete_back_from(oldest_notification_bst):
         deletion_date = deletion_date + dateutil.relativedelta.relativedelta(months=3)
 
     return deletion_date
+
+
+@notify_celery.task(name="delete_unneeded_notification_history_by_hour")
+def delete_unneeded_notification_history_by_hour():
+    # This task will delete all of the notification_history table older than 1 Jan 2023 BST
+    #
+    # This task will create lots of tasks, each one responsible for deleting a particular hour of
+    # notification_history that is no longer needed
+    #
+    # This retention limit is hardcoded and was originally picked from
+    # https://github.com/alphagov/notifications-aws/blob/main/decisions/2022-12-01-notification-history-retention-period.md
+    # It was supposed to be 2023-4-1.
+    # However, at the time of writing this code we realised the FtBillingLetterDispatch table has been introduced
+    # which means we need to store letters older than 2023-4-1 in order to rebuild that table (because that table
+    # uses the date of dispatch, not the date of creation for which date to bill for). To keep it simple, we keep
+    # an extra quarters worth of data giving us plenty of buffer
+    #
+    # In the future, we will be able to update this retention_limit value when we have progressed 3 quarters
+    # into the next financial year
+    #
+    # Arbitrary start_datetime, just slightly older than the oldest notification in the notification_history
+    # table at the time of writing
+    start_datetime = datetime(2020, 8, 1, 0, 0, 0)
+    retention_limit = datetime(2023, 1, 1, 0, 0, 0)
+
+    while start_datetime < retention_limit:
+        delete_unneeded_notification_history_for_specific_hour.apply_async(
```

Comment:
I wouldn't hate a log message here so that we can see a clear record of all the hours we're generating, just as an audit trail of what we've made and a clear trail that the start/stop is correct when we do a prod run.

Reply:
I've added this in and rebased it as part of my original commit